flink-libraries/flink-table/src/main/java/org/apache/flink/table/api/catalog/ReadableWritableCatalog.java

```diff
@@ -0,0 +1,175 @@
+package org.apache.flink.table.api.catalog;
+
+import org.apache.flink.table.api.TableAlreadyExistException;
+import org.apache.flink.table.api.TableNotExistException;
+import org.apache.flink.table.api.catalog.exceptions.DatabaseAlreadyExistException;
+import org.apache.flink.table.api.catalog.exceptions.DatabaseNotExistException;
+import org.apache.flink.table.plan.stats.TableStats;
+
+/**
+ * An interface responsible for manipulating catalog metadata.
+ */
+public interface ReadableWritableCatalog extends ReadableCatalog {
+
+	// ------ databases ------
+
+	/**
+	 * Adds a database to this catalog.
+	 *
+	 * @param dbName    	The name of the database to add.
+	 * @param database        The database to add.
+	 * @param ignoreIfExists Flag to specify behavior if a database with the given name already
+	 *                       exists: if set to false, it throws a DatabaseAlreadyExistException,
+	 *                       if set to true, nothing happens.
+	 * @throws DatabaseAlreadyExistException
+	 *                       thrown if the database does already exist in the catalog
+	 *                       and ignoreIfExists is false
+	 */
+	void createDatabase(String dbName, CatalogDatabase database, boolean ignoreIfExists)
+		throws DatabaseAlreadyExistException;
+
+	/**
+	 * Deletes a database from this catalog.
+	 *
+	 * @param dbName        Name of the database to delete.
+	 * @param ignoreIfNotExists Flag to specify behavior if the database does not exist:
+	 *                          if set to false, throw an exception,
+	 *                          if set to true, nothing happens.
+	 * @throws DatabaseNotExistException thrown if the database does not exist in the catalog
+	 */
+	void dropDatabase(String dbName, boolean ignoreIfNotExists) throws DatabaseNotExistException;
+
+	/**
+	 * Modifies an existing database.
+	 *
+	 * @param dbName        Name of the database to modify.
+	 * @param newDatabase           The new database to replace the existing database.
+	 * @param ignoreIfNotExists Flag to specify behavior if the database does not exist:
+	 *                          if set to false, throw an exception,
+	 *                          if set to true, nothing happens.
+	 * @throws DatabaseNotExistException thrown if the database does not exist in the catalog
+	 */
+	void alterDatabase(String dbName, CatalogDatabase newDatabase, boolean ignoreIfNotExists)
+		throws DatabaseNotExistException;
+
+	/**
+	 * Renames an existing database.
+	 *
+	 * @param dbName        Name of the database to modify.
+	 * @param newDbName     New name of the database.
+	 * @param ignoreIfNotExists Flag to specify behavior if the database does not exist:
+	 *                          if set to false, throw an exception,
+	 *                          if set to true, nothing happens.
+	 * @throws DatabaseNotExistException thrown if the database does not exist in the catalog
+	 */
+	void renameDatabase(String dbName, String newDbName, boolean ignoreIfNotExists)
```

Comment:
does it need to throw a 'DatabaseAlreadyExistsException' when the 'newDbName' is already taken?

Reply:
Well, DatabaseAlreadyExistsException is reserved for the case in which user tries to CREATE a db that's already existing, which is different from the case here. Potentially, we could create a new exception class for the case you mentioned, but I think a runtime exception might suffice for now.